### PR TITLE
keys(::Generator) pass through

### LINKED ISSUE #34674: keys(::Generator) for find* and arg*
Tl;Dr:
Add
```julia
Base.keys(gen::Base.Generator) = Base.keys(gen.iter)
```
to Base.

-----

Partially related to my comment in https://github.com/JuliaLang/julia/issues/34368#issuecomment-582365454 regarding the support of traits on generators.
Currently there is no way to use find* on generators. Reason is that there is no keys function:
```julia
julia> findmax(abs(x-1) for x in [1,5,2,9])
ERROR: MethodError: no method matching keys(::Base.Generator{Array{Int64,1},var"#97#98"})
```
But since generators are said to be the lazy variant of `map` and `map` preserves the indices it would be good to define that missing `keys` method.

That would allow for some very generic `findmax(iter)` and `argmax(iter)` as arbitrary lazy maps are allowed while maintaining good performance and low memory footprint.

Edit: adding the missing find* and arg* with custom transform functions would be appreciated aswell to allow for the convenient do-block notation.

### PR DESCRIPTION
**Pull request by rapus95**
Let keys pass through to allow lightweigth/lazy find* and arg* functions
Fixes https://github.com/JuliaLang/julia/issues/34674

### CODE REVIEW

**Review by aminya — APPROVED** [review]

**Comment by rapus95** [comment]
What would be the intended tests here?
is that already enough?
```
@test keys(x^2 for x in 1:10) == keys(1:10)
```

**Comment by KlausC** [comment]
As a test maybe more illustrative:
```
@test keys(x^2 for x in -1:0.5:1) == 1:5
```
I propose to put that into `test/functional.jl`

**Comment by StefanKarpinski** [comment]
Triage finds it a little weird but approves since it makes find work and doesn't seem harmful. Would be good to add @KlausC's test and a NEWS entry, however. A use case would also be good.

**Review by mbauman — DISMISSED** [review]

**Comment by rapus95** [comment]
@mbauman sure, got interrupted by something else after rebasing such an old branch for the first time 😇
For the NEWS entry, into which section should this go?

File: test/functional.jl [hunk 1]
```diff
@@ -134,6 +134,9 @@ let gen = ((x,y) for x in 1:10, y in 1:10 if x % 2 == 0 && y % 2 == 0),
     @test collect(gen) == collect(gen2)
 end
 
+# keys of a generator for find* and arg* (see #34678)
+@test keys(x^2 for x in -1:0.5:1) == 1:5
+
```
Comment (mbauman) [inline test/functional.jl:139]:
```suggestion
@test findall(!iszero, x^2 for x in -1:0.5:1) == [1, 2, 4, 5]
@test argmin(x^2 for x in -1:0.5:1) == 3

```

Let's also test your main motivation

**Comment by thofma** [comment]
Kind of fixes https://github.com/JuliaLang/julia/issues/23680

Instead of `indmin` I just have to use `findmin`.

**Comment by rapus95** [comment]
@thofma once this lands, you can just use `indmin` again since `keys` for `Generator` won't be missing anymore.

**Comment by rapus95** [comment]
@mbauman well since I guess 1.5 is already pinned now (branched away from master), but to not to add it to the wrong news file, this'll go into 1.6, won't it? (so I can add a news entry straight to the 1.6 news file)

I'd put it into the "standard library changes" section.

**Comment by tkf** [comment]
I think `keys`, `values` and `getindex` should satisfy the invariance

```julia
isequal(collect(values(A)), [A[k] for k in keys(A)])
```

So, I'm against this patch _unless_ `values` and `getindex` are defined for `Generator`.  See #36175.

**Comment by rapus95** [comment]
bump for 1.6 👀

**Comment by rapus95** [comment]
@fredrikekre please remove the "needs test" label. (Since there already is one). And can you add it to a 1.6 milestone? (so that it won't go unnoticed upon freezing, especially since triage already accepted the standalone and https://github.com/JuliaLang/julia/pull/37648 holds the `getindex` for generators)

**Comment by oscardssmith** [comment]
Rerunning co, then good to merge in my view.

**Review by oscardssmith — APPROVED** [review]

**Comment by oscardssmith** [comment]
@StefanKarpinski me closing and re-opening doesn't seem to have trigered CI. Can you make it re-run?

**Comment by rapus95** [comment]
> @StefanKarpinski me closing and re-opening doesn't seem to have trigered CI. Can you make it re-run?

@oscardssmith To me it seems like buildbot was triggered and currently is only missing tester_macos64 and tester_win32

**Comment by oscardssmith** [comment]
oh yeah. Sorry for the noise. It might be worth rebasing this onto master so we can get a fully green CI.

**Comment by musm** [comment]
This looks like it still has a triage label, was this approved? Typically, the triage  label would be removed before we allow a PR to be merged.

**Comment by oscardssmith** [comment]
Triage approved back in March. I think it was re-added to consider back-porting once merged, but I'm not sure.

**Comment by musm** [comment]
To be sure, it would be best to hold off then until triage-reapproves and once the requested review is approved.

**Comment by JeffBezanson** [comment]
Triage is ok with this again. I came around since it basically seems consistent with `axes`, which we already define here. We don't want to backport it but it can be merged on master.

**Comment by rapus95** [comment]
Ok that's fine, then we should be able to discuss the long-term meaning of Generator and thus, if we also want to include the other issues/PRs (#37648, #34368) which would lead to a better rounded feature addition in 1.7

**Comment by simeonschaub** [comment]
Thank you for the contribution! Sorry this took so long. In the future, also feel free to bump PRs every once in a while, because otherwise things sometimes get forgotten.

**Pull request merged**